File: services/amadeusHotels.js
const axios = require('axios');
const db = require('./db');
const path = require('path');

const MOCK_PATH = path.join(__dirname, '..', 'mock', 'mock_hotels.json');

class AmadeusHotels {
  constructor() {
    this.clientId = process.env.AMADEUS_CLIENT_ID;
    this.clientSecret = process.env.AMADEUS_CLIENT_SECRET;
    this.baseUrl = process.env.AMADEUS_ENV === 'production'
      ? 'https://api.amadeus.com/v1'
      : 'https://test.api.amadeus.com/v1';
    this.accessToken = null;
    this.tokenExpiry = null;
  }

  async getAccessToken() {
    if (this.accessToken && this.tokenExpiry && Date.now() < this.tokenExpiry) {
      return this.accessToken;
    }
    const res = await axios.post(
      `${this.baseUrl}/security/oauth2/token`,
      new URLSearchParams({
        grant_type: 'client_credentials',
        client_id: this.clientId,
        client_secret: this.clientSecret
      }),
      { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } }
    );
    this.accessToken = res.data.access_token;
    this.tokenExpiry = Date.now() + (res.data.expires_in - 300) * 1000;
    return this.accessToken;
  }

  // Try Amadeus; on any failure return mock results (so demo works)
  async searchHotels(params = {}) {
    const city = params.city || params.cityCode || params.q;
    if (!city) throw new Error('searchHotels requires a city');

    try {
      const token = await this.getAccessToken();
      const response = await axios.get(
        `${this.baseUrl}/v1/reference-data/locations/hotels/by-city`,
        {
          headers: { Authorization: `Bearer ${token}` },
          params: { cityCode: city }
        }
      );

      const hotels = response.data?.data || [];
      // Save only when we have real data
      if (hotels.length > 0) {
        await this.saveHotels(hotels, city);
      }
      return hotels.length > 0 ? hotels : this._loadMock(city);
    } catch (err) {
      console.error('Hotel search failed (using mock):', err.response?.data || err.message);
      return this._loadMock(city);
    }
  }

  _loadMock(city) {
    try {
      // lazy-load to avoid breaking if mock missing
      const mock = require(MOCK_PATH);
      return mock.filter(h => !city || h.city === city);
    } catch (e) {
      console.error('Failed to load mock hotels:', e.message);
      return [];
    }
  }

  async saveHotels(hotels, city) {
    const client = await db.pool.connect();
    try {
      await client.query('BEGIN');
      const providerRes = await client.query(
        `INSERT INTO providers (name, type, created_at)
         VALUES ($1,$2,now())
         ON CONFLICT (name) DO UPDATE SET type = EXCLUDED.type
         RETURNING id`,
        ['Amadeus', 'hotel']
      );
      const providerId = providerRes.rows[0].id;

      for (const h of hotels.slice(0, 10)) {
        await client.query(
          `INSERT INTO hotels
             (provider_id, name, city, address, stars, data, created_at)
           VALUES ($1,$2,$3,$4,$5,$6,now())
           ON CONFLICT (name) DO NOTHING`,
          [
            providerId,
            h.name || h.hotelName || 'Unknown',
            city,
            (h.address && h.address.lines) ? h.address.lines.join(', ') : (h.address?.line || null),
            h.rating ? parseInt(h.rating) : null,
            h
          ]
        );
      }

      await client.query('COMMIT');
    } catch (e) {
      await client.query('ROLLBACK');
      console.error('saveHotels error', e);
    } finally {
      client.release();
    }
  }
}

module.exports = new AmadeusHotels();

File: AI-Backend/modules/amadeus_hotels.py
import os
from amadeus import Client, ResponseError

_client = None


def _get_client() -> Client:
    global _client
    if _client is None:
        _client = Client(
            client_id=os.environ["AMADEUS_API_KEY"],
            client_secret=os.environ["AMADEUS_API_SECRET"],
        )
    return _client


def search_hotels(
    city: str,
    check_in_date: str,
    check_out_date: str | None = None,
    adults: int = 1,
) -> list[dict]:
    """Return up to 5 hotel offers for a city using the Amadeus Hotel Search API."""
    client = _get_client()
    try:
        # Step 1: get hotel IDs for the city
        hotel_list = client.reference_data.locations.hotels.by_city.get(cityCode=city.upper())
        hotel_ids = [h["hotelId"] for h in (hotel_list.data or [])[:20]]
        if not hotel_ids:
            return []
    except ResponseError as e:
        print(f"[amadeus_hotels] hotel list ResponseError: {e}")
        return []

    params = {
        "hotelIds": hotel_ids,
        "adults": adults,
        "checkInDate": check_in_date,
    }
    if check_out_date:
        params["checkOutDate"] = check_out_date

    try:
        response = client.shopping.hotel_offers_search.get(**params)
        return (response.data or [])[:5]
    except ResponseError as e:
        print(f"[amadeus_hotels] offers ResponseError: {e}")
        return []
